exporters/etw/include/opentelemetry/exporters/etw/etw_data.h

```diff
@@ -0,0 +1,536 @@
+  /// <returns>Tracer instance</returns>
+  Tracer(trace::TracerProvider &parent,
+         nostd::string_view providerId     = "",
+         ETWProvider::EventFormat encoding = ETWProvider::EventFormat::ETW_MANIFEST)
+      : trace::Tracer(),
+        parent(parent),
+        provId(providerId.data(), providerId.size()),
+        encoding(encoding)
+  {
+    provHandle = etwProvider().open(provId, encoding);
+  };
+
+  /// <summary>
+  /// Start Span
+  /// </summary>
+  /// <param name="name">Span name</param>
+  /// <param name="options">Span options</param>
+  /// <returns>Span</returns>
+  virtual nostd::shared_ptr<trace::Span> StartSpan(
+      nostd::string_view name,
+      const common::KeyValueIterable &attributes,
+      const trace::SpanContextKeyValueIterable &links,
+      const trace::StartSpanOptions &options = {}) noexcept override
+  {
+    // TODO: support attributes
+    UNREFERENCED_PARAMETER(attributes);
+    UNREFERENCED_PARAMETER(links);
+    return trace::to_span_ptr<Span, Tracer>(this, name, options);
+  };
+
+  /// <summary>
+  /// Force flush data to Tracer, spending up to given amount of microseconds to flush.
+  /// </summary>
+  /// <param name="timeout">Allow Tracer to drop data if timeout is reached</param>
+  /// <returns>void</returns>
+  virtual void ForceFlushWithMicroseconds(uint64_t) noexcept override{};
+
+  /// <summary>
+  /// Close tracer, spending up to given amount of microseconds to flush and close.
+  /// </summary>
+  /// <param name="timeout">Allow Tracer to drop data if timeout is reached</param>
+  /// <returns></returns>
+  virtual void CloseWithMicroseconds(uint64_t) noexcept override
+  {
+    etwProvider().close(provHandle);
+  };
+
+  /// <summary>
+  /// Add event data to span associated with tracer
+  /// </summary>
+  /// <param name="span"></param>
+  /// <param name="name"></param>
+  /// <param name="timestamp"></param>
+  /// <param name="attributes"></param>
+  /// <returns></returns>
+  void AddEvent(Span &span,
+                nostd::string_view name,
+                core::SystemTimestamp timestamp,
+                const common::KeyValueIterable &attributes) noexcept
+  {
+    // TODO: associate events with span
+    (void)span;
+    // TODO: respect original timestamp
+    (void)timestamp;
+
+    // Unfortunately we copy right now since we don't have the knowledge of original map object :-(
+    std::map<nostd::string_view, common::AttributeValue> m;
+    attributes.ForEachKeyValue([&](nostd::string_view key, common::AttributeValue value) noexcept {
+      m[key] = value;
+      return true;
+    });
+    m["name"] = name.data();
+
+#ifdef HAVE_TLD
+    if (encoding == ETWProvider::ETW_MANIFEST)
+    {
+      etwProvider().writeTld(provHandle, m);
+      return;
+    }
+#endif
+
+#ifdef HAVE_MSGPACK
+    if (encoding == ETWProvider::ETW_MSGPACK)
+    {
+      etwProvider().writeMsgPack(provHandle, m);
+      return;
+    }
+#endif
```

Comment:
You are right. We should. Because `write` abstracts away either / or. I think Mishal only tested the MsgPack path because for the other implementation we did not share the `TraceLoggingDynamic.h` in OSS yet.

Reply:
Remaining span data (trace id, span id, links, start time, and duration) must not be dropped, but need to be sent too.